kroxylicious/src/main/java/io/kroxylicious/proxy/codec/CorrelationManager.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.kroxylicious.proxy.codec;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Manages correlation ids for a single connection (across the proxy) between a single client
+ * and a single broker.
+ */
+public class CorrelationManager {
+
+    // TODO use a specialized map
+    /* test */ final Map<Integer, Correlation> brokerRequests = new HashMap<>();
+
+    private final AtomicInteger upstreamIds;
```

Comment:
Why atomic?
Atomic instruction are very expensive, especially.
If the Id for fast lookup gotta be shared (because of uniqueness constraints), I suggest to:
- use a generic shared SequenceGenerator (they can use an atomic long/int)
- have a thread/event-loop confined correlation manager, still using a single threaded specialized map for the id-correlation mapping
- eventually using a separate or part of the shared sequence generator's dispose method for "recycled" sequence ids: this can enable using expandable arrays ie just an ArrayList instead of a my map


Last point can be seen as premature optimization, so feel free to ignore me, but having a single threaded fast correlation lookup path, seems a good thing to me; making uses of a thread-per-core partitioning of resources is key to avoid using concurrency if not needed

Reply:
Yeah, I don't this atomicity is needed here, since this CorrelationManager can assume it is used for a single connection. If we ended up allowing filters to make requests via shared connections then we'd need a different `CorrelationManager` impl which had atomic allocation. But it's far from clear if we need that right now. I'll fix this.